### PR TITLE
Save submitted answers

### PR DESCRIPTION
**Pull request by JamesM25**

### CODE REVIEW

File: src/Index.ts [hunk 1]
```diff
@@ -30,46 +32,69 @@ function checkFields(body: Object, res: any, ...fields: string[]): boolean {
...
+        checkFields(req.body, res, "question")
+        return await clearHistory(db, req.body.question);
+    });
 });
 
-app.post('/answer', async (req, res) => {
-    if (!checkFields(req.body, res, "query", "question"))
-        return;
+app.use((err: Error, req: Request, res: Response, next: NextFunction) => {
+    //console.error(err.stack);
```
Comment (kellerflint) [inline src/Index.ts]: Either uncomment if intended to be part of logging or remove

File: src/service/AssignmentService.ts [hunk 2]
```diff
@@ -40,63 +51,79 @@ export async function getQuestionList(db: DatabaseConnection, assignmentId: numb
...
+
+    const uaResult = await db.exec(`SELECT *
+        FROM users_assignments
+        WHERE user_id = @user AND assignment_id = @assignment;`,
+        uaParams);
+
+    checkQueryResult(uaResult, false);
+    
+    if (uaResult.rows.length === 0) {
+        // TODO: Make sure there are no duplicates
```
Comment (kellerflint) [inline src/service/AssignmentService.ts]: I'd rather this be recorded as an issue or work item or something than leaving a comment it in the code

File: src/service/AssignmentService.ts [hunk 3]
```diff
@@ -108,11 +135,78 @@ export async function checkAnswer(db: DatabaseConnection, questionId: number, us
...
+            history: userQuery,
+            score: isCorrect ? question.points : 0
+        });
+
+        checkQueryResult(answerInsertResult, false);
+    } else {
+        // UPDATE existing answer
+        
+        const answerId = answerSelectResult.rows[0].id;
+        const history = answerSelectResult.rows[0].query_history.split(HISTORY_DELIMITOR);
```
Comment (kellerflint) [inline src/service/AssignmentService.ts]: I think this should be done as a separate QueryHistory table with an FK reference to the answer.

File: src/service/AssignmentService.ts [hunk 4]
```diff
@@ -108,11 +135,78 @@ export async function checkAnswer(db: DatabaseConnection, questionId: number, us
...
+            WHERE q.id = @id`,
+        {id: questionId, user: DEFAULT_USER_ID});
+    
+    checkQueryResult(uaResult, true);
+
+    const uaId = uaResult.rows[0].id;
+
+    const result = await(db.exec(`
+        UPDATE answers SET query_history = ''
+        WHERE ua_id = @ua AND question_id = @question`,
```
Comment (kellerflint) [inline src/service/AssignmentService.ts]: I'm not a fan of writing out SQL in these functions. I think the better apporach would be to use the repository pattern and handle the queries in a separate file that's sole job is to interact with the DB. SQL code should probably not appear in our AssignmentService, it should just be our business logic. Including data access details means we're violating the separation of concerns principle.

Ideally this clearHistory function for example would just look more like this:
clearHistory(...) { 
  const userAssignmentResult = await assignmentRepo.getAssignmentByQuestionId(questionId);
  checkQueryResults(userAssignmenetResult, true);
  const result = await assignmentRepo.setQueryHistory(questionId, uaId)
}

This is much better for other developers working on our code as we're working only at the abstraction level of the business rules. It's nice to only show what they need to know at this level, not to show them all the details of data access they probably don't care about. If they want those details, they can always click into the repo function.

**Review by kellerflint — CHANGES_REQUESTED** [review]
Overall looks great! Error handling is much better. I have some suggestions for data access and the QueryHistory thing we discussed the other day. Ping me when you have changes/questions.

File: src/service/AssignmentService.ts [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
 import DatabaseConnection from "./DatabaseService";
 import { QueryResult } from "../model/QueryResult";
 import RouteResult from "../model/RouteResult";
-import { STATUS_CODES } from "http";
+import AppError from "../model/AppError";
+import AssignmentRepository from "./AssignmentRepository";
+import QuestionAnswer from "../model/QuestionAnswer";
+import Question from "../model/Question";
+
+const HISTORY_DELIMITOR = '\0';
```
Comment (kellerflint) [inline src/service/AssignmentService.ts]: Is this still necessary?

**Review by kellerflint — DISMISSED** [review]

**Review by kellerflint — DISMISSED** [review]
👍

**Review by kellerflint — APPROVED** [review]
👍

**Review by jenniferzal — APPROVED** [review]

**Pull request merged**